### PR TITLE
Proficiencies UI

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

#### Purpose of change
This was brought to my attention in the comments for https://github.com/CleverRaven/Cataclysm-DDA/pull/56487#issuecomment-1084798832

The proficiencies list in the player info screen is getting pretty long and unwieldy. There should be a more convenient way to view the player's proficiencies.

#### Describe the solution

https://user-images.githubusercontent.com/12537966/161436691-6c5581d1-d620-4f82-ab79-c5528dda7fcf.mp4

The proficiencies UI can be accessed from the player info by pressing `p` or by pressing `Enter` while selecting a proficiency.

<hr>

##### Proficiency Categories

The current proficiencies have been categorized using the `"category"` field. A category corresponds to a `"proficiency_category"` type object:

```json
{
  "type": "proficiency_category",
  "id": "prof_firstaid",
  "name": "Health Care",
  "description": "Proficiencies for treating injuries and general health care knowledge."
}
```

<hr>

##### Proficiency Filtering

Pressing the `f`ilter key brings up the search prompt:

prof_filter

As the hint suggests, entering a string of text looks for proficiencies containing that string in the name. Several prefixes are included:
- `l:` - Returns learned proficiencies (i.e.: proficiencies with 100% progress)
ex: `l:alch` returns all learned proficiencies with "alch" somewhere in the name.
- `i:` - Same as above except for in-progress proficiencies (i.e.: <100% progress)
- `p:` - Returns all proficiencies with a progress rate above the given value. Must be a decimal number.
ex: `p:20.00` returns all proficiencies with at least 20% progress.

filter_text
filtered

<hr>

#### Describe alternatives you've considered
Various different layouts. I landed on this one as a nice balance between simplicity and information density.

prof_leads_to

#### Testing
See video and screenshots above.

#### Additional context
TODO:
- [x] Implement proficiency categories
- [x] Implement UI class and draw functions
- [x] Make UI accessible from player info
- [x] Implement filtering for proficiencies
- [x] Update documentation
- [x] Testing

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
This is above and beyond, great work!

**Comment by Night-Pryanik** [comment]
Please make this window wider, like in, for example, missions ui. Necessity to fold string consisting of 6 words (see Description line) feels like a problem with UI for me.

**Comment by dseguin** [comment]
> Please make this window wider, like in, for example, missions ui. Necessity to fold string consisting of 6 words (see Description line) feels like a problem with UI for me.

I adjusted the max width of the UI from 80 to 120 (I'd rather not increase it too much more because it's a fairly simple screen):

prof_ui

I also made sure that the window can scale down to much smaller screens (80x24):

prof_ui_80x24

**Comment by PatrikLundell** [comment]
Listing prerequisites for proficiencies you already know seems to be of limited use as it just shows you what you've already mastered, and thus no longer need to care about. Listing what you'd be able to proceed to would be much more useful, in particular if you'd be able to also look at proficiencies you haven't started learning.

As an example, I find the recipe for Foo, which relies on the proficiencies of Bar and Baz, but both of those seem to be advanced proficiencies, so I'd like to find out what I'd have to learn first.

This basically would turn the expansion menu into an information menu rather than a strictly PC menu. I'd default the expansion menu to either hide unstarted proficiencies, or list them at the end in a darker color.

File: src/proficiency_ui.cpp [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+shared_ptr_fast<ui_adaptor> _prof_window::create_or_get_ui_adaptor()
+{
+    shared_ptr_fast<ui_adaptor> current_ui = ui.lock();
+    if( !current_ui ) {
+        ui = current_ui = make_shared_fast<ui_adaptor>();
+        current_ui->on_screen_resize( [this]( ui_adaptor & cui ) {
+            init_ui_windows();
+            cui.position_from_window( catacurses::stdscr );
```
Comment (Qrox) [inline src/proficiency_ui.cpp:263]:
```suggestion
            cui.position_from_window( w_border );
```

If I understand your code correctly, `w_border` covers the entire window area, so you should use it as the argument of this call for the underlying UI to redraw correctly during resizing.

File: src/proficiency_ui.cpp [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    ctxt.register_directions();
+    ctxt.register_action( "HELP_KEYBINDINGS" );
+    ctxt.register_action( "PREV_CATEGORY" );
+    ctxt.register_action( "NEXT_CATEGORY" );
+    ctxt.register_action( "FILTER" );
+    ctxt.register_action( "QUIT" );
+
+    populate_categories();
+    shared_ptr_fast<ui_adaptor> current_ui = create_or_get_ui_adaptor();
+    init_ui_windows();
```
Comment (Qrox) [inline src/proficiency_ui.cpp:292]: There's no need to call it here because `ui_manager::redraw` will call it.

File: src/proficiency_ui.cpp [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+            std::string desc( _( "Available prefixes:" ) );
+            desc += "\n  ";
+            desc += _( "<color_c_light_cyan>l:</color>  Learned proficiencies" );
+            desc += "\n  ";
+            desc += _( "<color_c_light_cyan>i:</color>  In-progress proficiencies" );
+            desc += "\n  ";
+            desc += _( "<color_c_light_cyan>p:</color>  Returns proficiencies above the given progress.  "
+                       "ex: <color_c_yellow>p:95.00</color>" );
+            desc += "\n\n";
+            desc += _( "By default, the filter applies to proficiencies containing the given string."
```
Comment (Night-Pryanik) [inline src/proficiency_ui.cpp:334]:
```suggestion
            desc += _( "By default, the filter applies to proficiencies containing the given string. "
```

File: src/proficiency_ui.cpp (hunk 2 above)
Comment (dseguin) [reply src/proficiency_ui.cpp:292]:
```suggestion
```

File: src/proficiency_ui.cpp (hunk 3 above)
Comment (dseguin) [reply src/proficiency_ui.cpp:334]:
```suggestion
            desc += _( "By default, the filter applies to proficiencies containing the given string.  "
```
Cheers. I believe translations require 2 spaces after a dot, though.

File: src/proficiency_ui.cpp [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+            if( sel_prof >= static_cast<int>( cur_set.size() ) ) {
+                sel_prof = 0;
+            }
+        } else if( action == "LEFT" || action == "PREV_CATEGORY" ) {
+            current_cat--;
+            if( current_cat < 0 ) {
+                current_cat = std::max( static_cast<int>( cats.size() ) - 1, 0 );
+            }
+            sel_prof = 0;
+            top_prof = 0;
```
Comment (dseguin) [inline src/proficiency_ui.cpp:315]:
```suggestion
            if( filter_str.empty() ) {
                current_cat--;
                if( current_cat < 0 ) {
                    current_cat = std::max( static_cast<int>( cats.size() ) - 1, 0 );
                }
                sel_prof = 0;
                top_prof = 0;
            }
```

File: src/proficiency_ui.cpp [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        } else if( action == "RIGHT" || action == "NEXT_CATEGORY" ) {
+            current_cat++;
+            if( current_cat >= static_cast<int>( cats.size() ) ) {
+                current_cat = 0;
+            }
+            sel_prof = 0;
+            top_prof = 0;
```
Comment (dseguin) [inline src/proficiency_ui.cpp:322]:
```suggestion
            if( filter_str.empty() ) {
                current_cat++;
                if( current_cat >= static_cast<int>( cats.size() ) ) {
                    current_cat = 0;
                }
                sel_prof = 0;
                top_prof = 0;
            }
```

**Comment by dseguin** [comment]
@PatrikLundell I've added a field to display what each proficiency leads to:

prof_leads_to

**Comment by casswedson** [comment]
Can we have something like `can't learn` if time to learn here is 0
Screenshot from 2022-04-03 10-41-52

**Comment by dseguin** [comment]
> Can we have something like `can't learn` if time to learn here is 0

Like this?

cant_learn

**Comment by bananapheasant** [comment]
You. are. awesome!

**Pull request merged**